### ISSUE TITLE
instanceof type guard fails with a static `toString` method

### ISSUE THREAD

**Issue by chharvey** [microsoft/TypeScript#27276]
**TypeScript Version:**  3.0.3
**Options:**  `--strict --noImplicitAny --noImplicitThis --noImplicitReturns --target="ES2017" --module="CommonJS"`

**Search Terms:** instanceof type guard class "toString"

### Code:
```ts
declare class Thing { static toString(thing: any): string; } // <!-- here’s the problem
declare class Person extends Thing { name: string; }
declare class Robot extends Thing { serial: number; }

function test(smart_thing: Person|Robot): number {
	if (smart_thing instanceof Robot) {
		return smart_thing.serial // typeof `smart_thing` is Person|Robot,
		                          // and getting `.serial` fails
	} else return 0
}
```

### Expected behavior:
The type of `smart_thing` should be narrowed to `Robot`.

### Actual behavior:
The type of `smart_thing` has not been narrowed from `Person | Robot`.

When the static method `Thing.toString` is renamed or removed, or its parameter is deleted, the problem goes away.

### Playground Link:
[TypeScript Playground](https://www.typescriptlang.org/play/index.html#src=declare%20class%20Thing%20%7B%20static%20toString(thing%3A%20any)%3A%20string%3B%20%7D%0D%0Adeclare%20class%20Person%20extends%20Thing%20%7B%20name%3A%20string%3B%20%7D%0D%0Adeclare%20class%20Robot%20extends%20Thing%20%7B%20serial%3A%20number%3B%20%7D%0D%0A%0D%0Afunction%20test(smart_thing%3A%20Person%7CRobot)%3A%20number%20%7B%0D%0A%09if%20(smart_thing%20instanceof%20Robot)%20%7B%0D%0A%09%09return%20smart_thing.serial%20%2F%2F%20typeof%20%60smart_thing%60%20is%20Person%7CRobot%2C%0D%0A%09%09%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20and%20getting%20%60.serial%60%20fails%0D%0A%09%7D%20else%20return%200%0D%0A%7D%0D%0A)

**Comment by andy-ms** [comment]
What is the code in `Person.class.ts` and `Employee.class.ts`?

**Comment by chharvey** [comment]
@andy-ms typical class code. in this reduced example, it could be as simple as this:
```ts
// Person.class.ts
export default class Person {
	constructor() {}
}
```
```ts
// Employee.class.ts
import Person from './Person.class'
export default class Employee extends Person {
	id: number;
	constructor(id: number) {
		super()
		this.id = id
	}
}
```

**Comment by andy-ms** [comment]
That reduced example is working when I try it.

**Comment by chharvey** [comment]
@andy-ms Ok so I think I found part of the problem. It weirdly involves a static `toString` method that takes a parameter, that inherits into each member of the union. It has nothing to do with `import`s. I've updated my code in the OP and the Playground.

**Comment by thw0rted** [comment]
Hey, any movement on this?  I just hit the same issue and it took me quite a while to figure out.  I had an separate `fooToString` method that I moved into `Foo` as a `static toString` and suddenly all my type guards broke.  

I couldn't get the OP playground link to work for me, but you can paste this in to see what I mean:

````
class A {
    public foo: string;
}

class B implements A {
    public foo: string;
    public bar: number;

    static toString(b: B) { return b.bar; }
}

class C implements A {
    public foo: string;
    public baz: boolean;

    static toString2(c: C) { return c.baz; } 
}

let x: A;
if (x instanceof B) { console.log(x.bar); }  // .bar is an error because X is still typed as A, not B
if (x instanceof C) { console.log(x.baz); }  // .baz works fine
````

Note that it's not only an issue when the superclass has a static `toString`, it's also a problem when one of the subclasses (the one that you're trying to narrow *to*, if that makes sense) has it.

**Comment by jakebailey** [comment]
I'm looking at old bugs for ones that appear to be fixed; all of the snippets in this issue appear to have been fixed from at least 3.3 and up (maybe even older).
